---
title: "Tutorial for NetBIDShiny online server"
layout: default
nav_order: 3
permalink: /docs/tutorial4online
---

# Tutorial for NetBIDshiny online server

The purpose of NetBIDshiny online server: 

**provide pre-generated network files from TCGA, TARGET and GTEx for quick start of NetBID2 analysis and demo datasets for quick familiar of diverse types of NetBID2 result visualization**.

The public online version of NetBIDshiny can be found here [NetBIDshiny_runner](https://yulab-stjude.shinyapps.io/NetBID2_Runner/) and [NetBIDshiny_viewer](https://yulab-stjude.shinyapps.io/NetBID2_Viewer/). 

----------
## Quick Navigation

- [NetBIDshiny_runner: pre-generated network files and demo usage](#netbidshiny_runner-pre-generated-network-files-and-demo-usage)

- [NetBIDshiny_viewer: demo datasets and usage](#netbidshiny_viewer-demo-datasets-and-usage)

----------

## NetBIDshiny_runner: pre-generated network files and demo usage

### Pre-generated network files

#### GTEx

- Original transcriptome dataset from [GTEx](https://www.genome.gov/Funded-Programs-Projects/Genotype-Tissue-Expression-Project)

- We have pre-generated 96 networks from 48 tissues with each tissue has one TF network and one SIG network. 

- The original sample information could be obtained from [GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt](https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt)

- The original gene expression profile could be obtained from [GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz](https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz)

- Detailed network QC files could be found in [GTEx_network](tutorial4online/GTEx_network)

#### TARGET

- Original transcriptome dataset from [TARGET](https://portal.gdc.cancer.gov/)

- We have pre-generated 102 networks from 51 cell lines with each cell line has one TF network and one SIG network.

- Detailed network QC files could be found in [TARGET_network](tutorial4online/TARGET_network)


#### TCGA

- Original transcriptome dataset from [TCGA](https://portal.gdc.cancer.gov/)

- We have pre-generated 92 networks from 46 tumor types with each has one TF network and one SIG network.

- Detailed network QC files could be found in [TCGA_network](tutorial4online/TCGA_network)

### Demo usage

The mouse BPD (bronchopulmonary dysplasia) demo dataset from GEO database [GSE25286](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE25286), with network files generated by using normal human lung tissue from GTEx (human). The task was to find drivers in BPD vs normal. 

GO to online version [NetBIDshiny_runner](https://yulab-stjude.shinyapps.io/NetBID2_Runner/).  

![f1](f1.png)

Click the `CHOOSE THE RDATA FILE CONTAINING THE EXPRESSIONSET CLASS OBJECT` button and select the RData `demo_mouse_BPD_eset.RData`.

![f2](f2.png)

Click the `CHOOSE THE TF NETWORK FILE` button and select the network txt file with suffix `TF.txt` under `GTEx/` directory. 

![f3](f3.png)

Click the `CHOOSE THE SIG NETWORK FILE` button and select the network txt file with suffix `SIG.txt` under `GTEx/` directory. 

![f4](f4.png)

Click the `LOAD THE DATA` button and options will occur in the right panel.

![f5](f5.png)

Modify the options.

![f6](f6.png)

Input the `Project name` and the `START THE DRIVER ESTIMATION ANALYSIS` button will occur. 
Click the `START THE DRIVER ESTIMATION ANALYSIS` button.

![f7](f7.png)

Wait about 3-5mins, the web will turn to the result page. Click the `DOWNLOAD RESULT ZIP FILE` button to download the result zip file.

![f8](f8.png)

Unzip the file and all related results are in the directory. Here, the `analysis.par.Step.ms-tab.RData` file under `MR_result/{project_name}/DATA/` directory could be directly uploaded to the [NetBIDshiny_viewer](https://yulab-stjude.shinyapps.io/NetBID2_Viewer/) server for result visualization. 

## NetBIDshiny_viewer: demo datasets and usage

GO TO [NetBIDshiny_viewer](https://yulab-stjude.shinyapps.io/NetBID2_Viewer/).

### Default demo dataset.

-- the human MB (medulloblastoma) demo dataset from GEO database as in NetBID2: [GSE116028](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116028), with network files generated by using the same dataset. The task was to find drivers in Group4 (G4) vs others. 

When opening the server, the human MB demo dataset will be directly uploaded:

![f9](f9.png)

Wait until the following page occurs:

![f10](f10.png)

This is the main page for the NetBIDshiny viewer, which contains the following main parts (nearly the same to the interface of the NetBIDshiny local version).

Click on different tabs to draw more plots (we do not provide detailed tutorial here as the [Tutorial for viewer](docs/tutorial4viewer) also show this demo.).

### Other demo dataset.

The online server also contain another mouse BPD demo dataset for result visualization: 
-- the mouse BPD (bronchopulmonary dysplasia) demo dataset from GEO database [GSE25286](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE25286), with network files generated by using normal human lung tissue from GTEx (human). The task was to find drivers in BPD vs normal in P14. 

Click the `CHOOSE THE MASTER TABLE RDATA FILE` button, and a new window containing the demo dataset show up. Click to choose the `GSE25286_P14_BPD_vs_normal.RData`. Click `SELECT`.

![f11](f11.png)

Click the `LOAD/RELOAD THE UPLOADED RDATA`. And the volcano plot will shown, which quickly identify the top differentially expressed/activated drivers. Since the differential expression (DE)/activity (DA) statistics are calculated from certain comparisons, users need to choose which comparison to visualize. Users can also adjust the threshold for P-values, logFC and driver's target size.

![f12](f12.png)

Click the tab `NETBID_PLOT`, and click the `DRAW NETBID PLOT` button and the NETBID plot will shown, which get the statistics of the top differentially expressed/activated drivers. It creates two side-by-side heatmaps to show the result of NetBID analysis. Both the differential expression analysis (the right heatmap) and differential activity analysis (the left heatmap) of top drivers are shown in the plot.

![f13](f13.png)

Click the tab `GSEA_PLOT`, and click the `DRAW GSEA PLOT` button and the GSEA plot will shown, which get the detailed statistics of the top differentially expressed/activated drivers. Such as number of target genes, ranking of target genes in differential expressed file, differential expression (DE) and differential activity (DA) values.

![f14](f14.png)

Click the tab `HEATMAP`, and click the `DRAW HEATMAP` button and the heatmap will shown, which get the expression/activity pattern of the top drivers across all samples.

![f15](f15.png)

Click the tab `FUNCTION_ENRICH_PLOT`, and click the `DRAW FUNCTION ENRICH PLOT` button and the netbid plot will shown, which get the functional annotation for the top drivers. It uses the annotation from MSigDB database. User can choose multiple categories of gene sets and related statistics for calculation. As long as the main category is selected, all the gene sets from that category will be used, regardless of the selection of sub-category. For example, if user choose the 'C5:GO' main category, all 'BP', 'MF', 'CC' will be used. It draws a cluster plot based on binary matrix, to visualize the existence of genes in the enriched gene sets. The P-value of enrichment is also displayed in the plot.

![f16](f16.png)

Click the tab `BUBBLE_PLOT`, and click the `DRAW BUBBLE PLOT` button and the bubble plot will shown, which get the functional annotation for the top drivers and their target genes. It uses the annotation from MSigDB database. It combines the matrix bubble chart and the heat map, using bubble color to compare P-values (performed by Fisher's Exact Test) and bubble size to compare the intersected size for target genes. Rows are enriched gene set, columns are top drivers. Users can also check number of protein-coding genes targetted by each driver. The plot creation will take some time (around several seconds). The figure size could be very large and inconvenient to see. We suggest to download it as PDF.

![f17](f17.png)

Click the tab `TARGET_NET`, and click the `DRAW TARGETNET PLOT` button and the target network plot will shown, which show the sub-network structure of one selected driver or a pair of selected drivers. Edges of positively-regulated target genes are orange, edges of negatively-regulated target genes are green. The width of the edges shows the strength of regulation.

![f18](f18.png)

Click the tab `TARGET_FUNCTION_ENRICH_PLOT`, and click the `DRAW TARGET FUNCTION ENRICH PLOT` button and the function enrichment plot for the selected driver will shown, which get the functional annotation for the driver's target genes. 

![f19](f19.png)

Click the tab `CATEGORY_BOXPLOT`, and click the `DRAW CATEGORY BOXPLOT` button and the boxplot will shown, which get the distribution of one driver's expression/activity value across group samples. 

![f20](f20.png)

Click the tab `DRIVER_GSEA_PLOT`, and click the `DRAW DRIVER GSEA PLOT` button and the GSEA plot for one selected driver will shown. 

![f21](f21.png)

P.S:  Bottom right is the plot panel. It includes adjusting bars and figure download button. Users can save the plot as PNG by right click, or click the button to download it as PDF.








